File: src/main/java/discretemath/bool/expression/forms/ProductOfSumExpression.java
package discretemath.bool.expression.forms;

import java.util.Iterator;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;

import discretemath.bool.expression.BooleanExpression;
import discretemath.bool.expression.atomic.BooleanVariable;
import discretemath.bool.expression.atomic.BooleanVariables;
import discretemath.bool.expression.exception.VariableValueNotSpecifiedException;
import discretemath.bool.operator.FunctionallyCompleteOperatorsSet;
import discretemath.bool.operator.Operators;
import discretemath.bool.table.FunctionValue;
import discretemath.bool.table.TruthTable;
import discretemath.bool.table.TruthTableRow;

public class ProductOfSumExpression implements BooleanExpression {

	private final List<BooleanVariable> variables;

	private final LinkedHashSet<Maxterm> maxterms;

	private ProductOfSumExpression(List<BooleanVariable> variables, LinkedHashSet<Maxterm> maxterms) {
		this.variables = variables;
		this.maxterms = maxterms;
	}

	public static ProductOfSumExpression fromTruthTable(TruthTable truthTable) {
		int functionsCount = truthTable.getFunctionsCount();
		if (functionsCount != 1) {
			throw new IllegalArgumentException("Truth table must be contain exactly one function");
		}

		int variablesCount = truthTable.getVariablesCount();

		List<BooleanVariable> variables = BooleanVariables.getVariables(variablesCount);

		LinkedHashSet<Maxterm> maxterms = new LinkedHashSet<>();
		for (TruthTableRow truthTableRow : truthTable) {
			if (truthTableRow.getFunctionValue(0) == FunctionValue.FALSE) {

				LinkedHashSet<Literal> literals = new LinkedHashSet<>();
				for (int i = 0; i < truthTableRow.argumentsCount(); i++) {
					boolean value = truthTableRow.getArgument(i);
					BooleanVariable variable = variables.get(i);
					literals.add(new SimpleLiteral(variable, !value));
				}

				maxterms.add(new SimpleMaxterm(literals));
			}
		}

		return new ProductOfSumExpression(variables, maxterms);
	}

	public List<BooleanVariable> getVariables() {
		return variables;
	}

	public LinkedHashSet<Maxterm> getMaxterms() {
		return maxterms;
	}

	@Override
	public boolean evaluate(Map<BooleanVariable, Boolean> arguments) throws VariableValueNotSpecifiedException {
		boolean[] values = new boolean[maxterms.size()];

		Iterator<Maxterm> maxtermIterator = maxterms.iterator();

		for (int i = 0; i < maxterms.size(); i++) {
			Maxterm maxterm = maxtermIterator.next();

			values[i] = maxterm.evaluate(arguments);
		}

		return Operators.AND.evaluate(values);
	}

	@Override
	public BooleanExpression expressViaOperators(FunctionallyCompleteOperatorsSet operatorsSet) {
		throw new UnsupportedOperationException(); //TODO
	}

	@Override
	public BooleanExpression normalize() {
		throw new UnsupportedOperationException(); //TODO
	}

	@Override
	public String toString() {
		StringBuilder builder = new StringBuilder();
		for (Maxterm maxterm : maxterms) {
			builder.append("(").append(maxterm).append(")");
		}

		return builder.toString();
	}
}

File: src/main/scala/discretemath/bool/expression/forms/ProductOfSumExpression.scala
package discretemath.bool.expression.forms

import discretemath.bool.expression.BooleanExpression
import discretemath.bool.expression.atomic.BooleanVariable
import discretemath.bool.expression.exception.VariableValueNotSpecifiedException
import discretemath.bool.operator.FunctionallyCompleteOperatorsSet
import discretemath.bool.operator.Operators.AND
import discretemath.bool.table.{FunctionValue, TruthTable}

import scala.collection.immutable.ListSet

class ProductOfSumExpression private(val variables: Seq[BooleanVariable], val maxterms: ListSet[Maxterm]) extends BooleanExpression {

  override def getVariables: Set[BooleanVariable] = maxterms.flatMap(_.getVariables)

  @throws[VariableValueNotSpecifiedException]
  override def evaluate(arguments: Map[BooleanVariable, Boolean]): Boolean = {
    val functionValuesCount = 1 << variables.length

    return maxterms.size match {
      case 0 => true
      case `functionValuesCount` => false
      case _ => AND.evaluate(maxterms.view.map(_.evaluate(arguments)).toSeq)
    }
  }

  override def expressViaOperators(operatorsSet: FunctionallyCompleteOperatorsSet): BooleanExpression = AND.injectUsing(operatorsSet, maxterms.toSeq)

  override def minimize: BooleanExpression = ???

  override def isSatisfiable: Boolean = {
    val functionValuesCount = 1 << variables.length

    return maxterms.size match {
      case `functionValuesCount` => false
      case _ => true
    }
  }

  override def toString: String = maxterms.view.map(maxterm => "(" + maxterm + ")").mkString("")
}

object ProductOfSumExpression {

  def apply(truthTable: TruthTable): ProductOfSumExpression = {
    val variables = BooleanVariable.getN(truthTable.getVariablesCount)
    return apply(truthTable, variables)
  }

  def apply(truthTable: TruthTable, variables: IndexedSeq[BooleanVariable]): ProductOfSumExpression = {
    val functionsCount = truthTable.getFunctionsCount
    if (functionsCount != 1) {
      throw new IllegalArgumentException("Truth table must be contain exactly one function")
    }

    val variablesCount = truthTable.getVariablesCount
    assert(variables.length == variablesCount, s"Variable list size expected: $variablesCount but was ${variables.length}")

    val maxterms: ListSet[Maxterm] = truthTable.view
      .filter(row => row.getFunctionValue(0) == FunctionValue.FALSE)
      .map(row => {
        val literals: ListSet[Literal] = variables.view
          .zip(row.arguments)
          .map(tuple => Literal(tuple._1, !tuple._2))
          .to(ListSet)

        Maxterm(literals)
      }).to(ListSet)

    return new ProductOfSumExpression(variables, maxterms)
  }
}